src/plugins/intel_gpu/src/graph/loop.cpp

```diff
@@ -540,6 +540,23 @@ void loop_inst::preprocess_backedge_memory() {
             GPU_DEBUG_LOG << idx << ") add back_edge mapping with SINGLE_SHARED type, backedge_mem("
                             << backedge_mem << "), initial_mem(" << initial_mem << ")" << std::endl;
         }
+        // Add _exec_deps for backedge primitives
+        for (auto user : backedge_to_prim->get_node().get_users()) {
+            cldnn::primitive_id user_primitive_id;
+            if (user->can_be_optimized()) {
+                // When user is optimized out, use user's user.
+                user_primitive_id = user->get_users().front()->get_primitive()->id;
+            } else {
+                user_primitive_id = user->get_primitive()->id;
+            }
+            auto user_primitive = body_network->get_primitive(user_primitive_id);
+
+            if (std::find(backedge_from_prim->_exec_deps.begin(), backedge_from_prim->_exec_deps.end(), user_primitive)
+               != backedge_from_prim->_exec_deps.end()) {
+                backedge_from_prim->_exec_dep_ids.push_back(user_primitive_id);
+                backedge_from_prim->_exec_deps.push_back(user_primitive);
+               }
+        }
```

Comment:
There might be a case where deadlock can happens. e.g., 
```
input(backedge_to) ------> op1 ----->
                        />
                     op2 ----->result (backedge_from)
```
                              
  Here if result node is opt out, then it will be  either of the following case:
  1)  op2 becomes the new backedge_from
  => op2 and op1 will have cyclic dependency 
  2) or, still the result is backedge but using op2's memory
  => above change won't be applied for op2 => original issue still happen 
  
So I think we should do additional thing for this purpose, i.e., we need to prevent network output node (reorder) 
 from being optimized out, if the output is in the loop body network && it is the backedge of mem sharing

Reply:
Updated the condition in remove_redundant_reorder pass. Thanks!